### DISCUSSION TITLE
Implementing UI - resources?

### DISCUSSION THREAD

**Discussion by Ananym** [wixtoolset/issues#8862] [Questions]
I've been trying at length to put together a very rudimentary UI for my installer in wix v5, but I can't find any resources in this direction and have been failing miserably.

Are there any examples of wix v5 installers with UI adjustments available anywhere?
I've been trying to follow the instructions at https://wixtoolset.org/docs/tools/wixext/wixui/#customizing-a-dialog-set to take wix ui minimal, remove/skip a dialog, and add a custom dialog - but without a bigger picture understanding I'm just not getting anywhere, plagued with errors about not having dialogs available, overriding actions, missing icons and fonts, no matter how I try to juggle it.

**Comment by chrpai** [comment]
I have a FOSS project called IsWiX that does this. It has project templates that demonstrate bootstrapper UI and MSI Native UI.  The Native UI has a custom dialog that gets lit up by uncommenting a single line.  Google iswix tutorials or email me for a demo.

**Comment by wmanning** [comment]
There is two install types in wixtoolset:
- MSI with it's MSI UI which is dated 
- burn with a more modern UI.  Basically burn installs MSI, etc silently (aka not using the MSI UI) while presenting it's own UI.

Might get better help if you are clear.  Seems you have a MSI only and want to use the MSI UI?

**Comment by Ananym** [comment]
> I have a FOSS project called IsWiX that does this. 

Thanks! The native templates helped me work a couple of things out.

> Seems you have a MSI only and want to use the MSI UI?

To be perfectly honest I hadn't even worked out that nobody was using regular MSI UIs anymore.  Whenever I've seen Burn mentioned it's only as a tool for bundling rather than for UI replacement!  I'll definitely look into Burn UIs in the longer run.

In the meantime I'd still like to figure the basic MSI UI out if at all possible.  What I wasn't able to work out from those linked templates is how to hide/skip an existing dialog, or how to set which dialog is shown first -

If I try something like...
```
      <InstallUISequence>
                <Show Dialog="WelcomeDlg" Before="WelcomeEulaDlg" Condition="0" />
      </InstallUISequence>
```

Taking the original line and setting condition to 0, in the hopes that it will never be shown -
without the "override" qualifier, building errors with `The action 'WelcomeDlg' conflicts with a virtual 
symbol with the same id.`
with the override qualifier, building errors with `Duplicate WixAction with identifier 'InstallUISequence/WelcomeDlg' found. `
and I've not yet figured out this impasse - maybe this is entirely the wrong approach.

**Reply by wmanning** [reply]
Nothing wrong with using the MSI UI. Didn't mean to imply no one is using it; only that it's a bit dated.  I really just wanted to point out the differences in case you didn't realize. 

If you don't have anything to bootstrap (such as dotnet) and can work with some limitations of the MSI UI (note: I only use burn and only hear of some limitations from time to time, no experience myself) then I think it's a good way to go.  You only have one thing to deal with: a MSI, instead of both a MSI and a bundle (burn).

**Comment by chrpai** [comment]
@wmanning there is a third type (and fourth actually) of UI. MSI can be launched with a UI handler and it can use an embedded UI handler.   InstallScript MSI (which I dislike)  is probably the biggest example of this.  In this scenario the handler completely takes over the InstallUISequence and is responsible for running all steps.  It then transfers control to the InstallExecuteSequence and handles all the progress messages pumpbed back to it.  Finally it takes over again for  Install Interrupted, Error and Success scenarios.

**Reply by wmanning** [reply]
Thanks, didn't know that.

**Comment by chrpai** [comment]
@Ananym  I wouldn't say that no one is using Native MSI UI anymore.

Only 1 dialog gets sequenced in the InstallUISequence.  All others are brought into the wizard loop via control events (publish elements).

If you aren't already using ORCA it's a must have.  Install the Windows Platform SDK and then install the Orca MSI.  This is a tool that lets you edit a built database and look at the table data.   Then take a look at the Windows Installer SDK documentation specifically

https://learn.microsoft.com/en-us/windows/win32/msi/controlevent-table

There's a lot of 101 ground to cover hence why I offer free demos.   It's a lot easier to learn WiX/MSI if you have at least a minimal context to start with.
